dll/win32/advapi32/reg/hkcr.c

```diff
@@ -107,7 +118,10 @@ GetFallbackHKCRKey(
     /* Get the key name */
     ErrorCode = GetKeyName(hKey, &KeyName);
     if (ErrorCode != ERROR_SUCCESS)
+    {
+        *MachineKey = hKey;
```

Comment:
Why is it necessary to still return the hKey in MachineKey, even in case of error? (and same question for the GetPreferred* function below)

Reply:
Because there are places in the code that does stuff like this (and the pointed to HKEY was never initialized)

```
    ErrorCode = GetFallbackHKCRKey(hKey, &FallbackKey, FALSE);
    if (ErrorCode != ERROR_SUCCESS)
    {
        if (PreferredKey != hKey)
            RegCloseKey(PreferredKey);
```

Reply:
Hmm.... the problem is that there seems to be other failure paths in this function (and in the GetPrefered* one) that still do not initialize the *MachineKey on return :thinking:

Reply:
Probably. Can `GetKeySam` fail? I don't know. `RtlPrefixUnicodeString` "failing" was the important bit since that can happen often.

I would assume all of this needs to be rewritten when wow64 support gets added since it needs to do a lot more inspection of which keys need handling.

Reply:
Well, there is a failure path also for there, so there is no question to ask: you need to return something there too for consistency. If GetKeySam couldn't fail, then it wouldn't return any error code there and there wouldn't be any associated failure check.

Reply:
> so there is no question to ask

In a standalone call, sure. But in this function, it uses the handle higher up in the code so we know the handle is valid when it gets to this point.